File: app/audio/vad.py
"""基于 webrtcvad 的断句器：能量门 + 语音起止 + 句尾静音 + 超长强切。"""
from __future__ import annotations

import webrtcvad


class VadSegmenter:
    def __init__(
        self,
        sample_rate: int = 16000,
        frame_ms: int = 20,
        silence_ms: int = 600,
        aggressiveness: int = 2,
        max_segment_s: int = 15,
    ) -> None:
        self.vad = webrtcvad.Vad(aggressiveness)
        self.sample_rate = sample_rate
        self.frame_ms = frame_ms
        # 校验：webrtcvad 仅支持 8k/16k/32k/48k 且 10/20/30ms
        assert frame_ms in (10, 20, 30), "frame_ms 仅支持 10/20/30"

        self.silence_thresh = max(1, silence_ms // frame_ms)   # 句尾静音帧数
        self.start_thresh = 3                                   # 起句连续语音帧数
        self.max_frames = int(max_segment_s * 1000 / frame_ms)  # 超长强切

        self.in_speech = False
        self.silent_run = 0
        self.speech_run = 0
        self.seg_frames = 0

    def is_voice(self, frame: bytes) -> bool:
        try:
            return self.vad.is_speech(frame, self.sample_rate)
        except Exception:
            return False  # 静音/异常帧视为非语音

    def process(self, frame: bytes) -> str | None:
        """处理一帧，返回 'start' / 'end' / None。"""
        voice = self.is_voice(frame)
        if not self.in_speech:
            if voice:
                self.speech_run += 1
                if self.speech_run >= self.start_thresh:
                    self.in_speech = True
                    self.silent_run = 0
                    self.seg_frames = self.speech_run
                    return "start"
            else:
                self.speech_run = 0
        else:
            self.seg_frames += 1
            if voice:
                self.silent_run = 0
            else:
                self.silent_run += 1
            if self.silent_run >= self.silence_thresh or self.seg_frames >= self.max_frames:
                self.in_speech = False
                self.silent_run = 0
                self.speech_run = 0
                return "end"
        return None

    def reset(self) -> None:
        self.in_speech = False
        self.silent_run = 0
        self.speech_run = 0
        self.seg_frames = 0

File: go/internal/audio/vad.go
package audio

import (
	"fmt"

	"github.com/maxhawkins/go-webrtcvad"
)

// VadSegmenter 基于 webrtcvad 的断句器：语音起止 + 句尾静音 + 超长强切。
// 与 Python 版 app/audio/vad.py 使用同一个 webrtcvad C 库，阈值语义完全一致。
type VadSegmenter struct {
	vad        *webrtcvad.VAD
	sampleRate int
	frameMS   int

	silenceThresh int // 句尾静音帧数
	startThresh   int // 起句连续语音帧数
	maxFrames     int // 超长强切

	inSpeech  bool
	silentRun int
	speechRun int
	segFrames int
}

// NewVadSegmenter 创建断句器。frameMS 仅支持 10/20/30（webrtcvad 限制）。
func NewVadSegmenter(sampleRate, frameMS, silenceMS, aggressiveness, maxSegmentS int) (*VadSegmenter, error) {
	if frameMS != 10 && frameMS != 20 && frameMS != 30 {
		return nil, fmt.Errorf("frame_ms 仅支持 10/20/30，当前 %d", frameMS)
	}
	if aggressiveness < 0 || aggressiveness > 3 {
		aggressiveness = 2
	}
	v, err := webrtcvad.New()
	if err != nil {
		return nil, fmt.Errorf("初始化 webrtcvad 失败: %w", err)
	}
	if err := v.SetMode(aggressiveness); err != nil {
		return nil, fmt.Errorf("设置 VAD 模式失败: %w", err)
	}

	silenceThresh := silenceMS / frameMS
	if silenceThresh < 1 {
		silenceThresh = 1
	}
	return &VadSegmenter{
		vad:           v,
		sampleRate:    sampleRate,
		frameMS:       frameMS,
		silenceThresh: silenceThresh,
		startThresh:   3,
		maxFrames:     maxSegmentS * 1000 / frameMS,
	}, nil
}

// Event 断句事件。
type Event int

const (
	EventNone Event = iota
	EventStart
	EventEnd
)

// Process 处理一帧，返回断句事件。
func (s *VadSegmenter) Process(frame []byte) Event {
	voice := s.isVoice(frame)
	if !s.inSpeech {
		if voice {
			s.speechRun++
			if s.speechRun >= s.startThresh {
				s.inSpeech = true
				s.silentRun = 0
				s.segFrames = s.speechRun
				return EventStart
			}
		} else {
			s.speechRun = 0
		}
		return EventNone
	}

	s.segFrames++
	if voice {
		s.silentRun = 0
	} else {
		s.silentRun++
	}
	if s.silentRun >= s.silenceThresh || s.segFrames >= s.maxFrames {
		s.inSpeech = false
		s.silentRun = 0
		s.speechRun = 0
		return EventEnd
	}
	return EventNone
}

// Reset 重置状态。
func (s *VadSegmenter) Reset() {
	s.inSpeech = false
	s.silentRun = 0
	s.speechRun = 0
	s.segFrames = 0
}

// isVoice 静音/异常帧视为非语音（与 Python 版 try/except 行为一致）。
func (s *VadSegmenter) isVoice(frame []byte) bool {
	ok, err := s.vad.Process(s.sampleRate, frame)
	if err != nil {
		return false
	}
	return ok
}
